pulsar/internal/auth/oauth2.go

```diff
@@ -0,0 +1,153 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package auth
+
+import (
+	"crypto/tls"
+	"fmt"
+
+	"github.com/apache/pulsar-client-go/oauth2"
+	"github.com/apache/pulsar-client-go/oauth2/cache"
+	"github.com/apache/pulsar-client-go/oauth2/store"
+	"k8s.io/utils/clock"
+)
+
+const (
+	ConfigParamType                  = "type"
+	ConfigParamTypeClientCredentials = "client_credentials"
+	ConfigParamIssuerURL             = "issuerUrl"
+	ConfigParamAudience              = "audience"
+	ConfigParamKeyFile               = "privateKey"
+	ConfigParamClientID              = "clientId"
+)
+
+type oauth2AuthProvider struct {
+	clock  clock.Clock
+	issuer oauth2.Issuer
+	store  store.Store
+	source cache.CachingTokenSource
+}
+
+// NewAuthenticationOAuth2WithParams return a interface of Provider with string map.
+func NewAuthenticationOAuth2WithParams(params map[string]string) (Provider, error) {
+	issuer := oauth2.Issuer{
+		IssuerEndpoint: params[ConfigParamIssuerURL],
+		ClientID:       params[ConfigParamClientID],
+		Audience:       params[ConfigParamAudience],
+	}
+
+	// initialize a store of authorization grants
+	st := store.NewMemoryStore()
+
+	oauth2Type := params[ConfigParamType]
+	keyFile := params[ConfigParamKeyFile]
+	return NewAuthenticationOAuth2(issuer, st, oauth2Type, keyFile)
+}
+
+func NewAuthenticationOAuth2(
+	issuer oauth2.Issuer,
+	store store.Store,
+	oauth2Type, keyFilePath string) (Provider, error) {
```

Comment:
I see that you changed this constructor from `NewAuthenticationOAuth2(issuer, store)` to `NewAuthenticationOAuth2(issuer, store, oauth2type, keyfile)`.   Let me explain why I think that's not a good change.

The authentication provider requires an authorization grant to be able to obtain access tokens over time.  The grant is a long-lived credential, such as a keyfile or a refresh token.  In all cases, the +application+ obtains the grant beforehand and then stores it in a store.  The store might be persistent.  Later, the application creates an authorization provider with the store as a parameter.  The provider loads the grant from the store, and uses it over time to obtain access tokens.

I am trying to explain that the provider is not responsible for obtaining or storing new grants.  Hence the signature `NewAuthenticationOAuth2(issuer, store)`. 

Now we discuss the special case of initializing the provider from parameters - `NewAuthenticationOAuth2WithParams`.  One could imagine a combination of parameters that would open a pre-existing store but this would require more design work.  A simple scenario that is possible now is the keyfile scenario in combination with an in-memory store (where the access token would be discarded when the process terminates).  This scenario is implemented above.

Would you revert the change to this method that is seen in https://github.com/apache/pulsar-client-go/pull/313/commits/97d98f753995fc7879e69472efec6fb698ebed99?